### PR TITLE
fix(iOS): fullscreenmodal color scheme adaptability

### LINKED ISSUE #2002: React Native `useColorScheme()` hook stops working when using presentation: 'fullScreenModal'
### Description

Current behavior
I am trying to present a Stack.Navigator as a fullScreenModal of a different stack, like so:

```
<Stack.Navigator
  initialRouteName="Splash"
  screenOptions={{
    ...GlobalStackOptions,
    headerTintColor: design.primaryText as string,
  }}>
  {/* Modal flows */}
  <Stack.Group
    screenOptions={{presentation: 'fullScreenModal', headerShown: false}}>
    <Stack.Screen name="MyScreen" component={ScreenComponent} />
  </Stack.Group>
</Stack.Navigator>
```

My app needs to support light and dark mode, which I am doing by utilizing React Native's useColorScheme() hook.
The rest of my app responds instantly to changes in the user's color scheme preference, but when a screen is in a fullScreenModal context, useColorScheme stops returning a correct value when the device color scheme changes.

Using formSheet works properly.

fullScreenModal:
https://github.com/react-navigation/react-navigation/assets/66689851/8931a780-3ae2-42ca-8ced-d968ae25cd23

formSheet:
https://github.com/react-navigation/react-navigation/assets/66689851/1d3740da-70a1-4b4b-9931-a01aa159beba

Expected behavior
I expect the UI to respond properly to the system color scheme when a screen is presented via a fullScreenModal.

### Steps to reproduce

1. add this to a fullScreenModal screen:

```
  const scheme: ColorSchemeName = useColorScheme();
  console.log('scheme: ', scheme);

```
2. Open that screen in your app
3. Toggle the system appearance using one of several methods (xcode menu, settings etc)
4. See the console log is not triggered
5. Do this with any other screen type like a formSheet or a regular push screen and you'll see the value gets toggled properly 

### Snack or a link to a repository

https://snack.expo.dev/w8orUoZon

Note I'm getting an error in snack but you should be able to use this code to check on a real device, following the Steps above.

### Screens version

3.20.0

### React Native version

0.72.3

### Platforms

iOS

### JavaScript runtime

Hermes

### Workflow

React Native (without Expo)

### Architecture

None

### Build type

Release mode

### Device

Real device

### Device model

IOS 17 iPhone 13 Pro

### Acknowledgements

Yes

### PR DESCRIPTION
**Pull request by alduzy**
## Description

This PR fixes the color scheme adaptability when using `presentation: fullScreenModal`.
When changing the appearance mode between `light` and `dark` the screens reacted immediately except for the fullScreenModal presentation style.

When using a `UIModalPresentationFullScreen` the views belonging to the presenting view controller are removed after the presentation completes. After investigating the view hierarchy for a screen with `presentation: fullScreenModal` it turned out that the `RCTRootView` is removed, thus the `traitCollectionDidChange` observer ([see here](https://github.com/facebook/react-native/blob/d3e0430deac573fd44792e6005d5de20e9ad2797/packages/react-native/React/Base/RCTRootView.m#L362)) is not working anymore.

The solution was to add an extra observer for `RNSScreenStackPresentationFullScreenModal` to be able to send proper notifications when no RCTRootView is present. As the observer is working synchronously it's usage is limited for this presentation style only.

Fixes #2002.

## Changes

- added repro `Test2002.tsx`
- added `traitCollectionDidChange` observer for fullScreenModal

### Before

https://github.com/software-mansion/react-native-screens/assets/91994767/52fa4e92-3baa-49e2-b278-7be57c4d28b3

### After
https://github.com/software-mansion/react-native-screens/assets/91994767/74c62c45-c793-4f63-81fc-d68d4000fea6

## Test code and steps to reproduce

- added `Test2002.tsx` to test examples

## Checklist

- [ ] Included code example that can be used to test this change
- [ ] Updated TS types
- [ ] Updated documentation: 
  - [ ] https://github.com/software-mansion/react-native-screens/blob/main/guides/GUIDE_FOR_LIBRARY_AUTHORS.md
  - [ ] https://github.com/software-mansion/react-native-screens/blob/main/native-stack/README.md
  - [ ] https://github.com/software-mansion/react-native-screens/blob/main/src/types.tsx
  - [ ] https://github.com/software-mansion/react-native-screens/blob/main/src/native-stack/types.tsx
- [ ] Ensured that CI passes

### CODE REVIEW

File: ios/RNSModalScreen.mm [hunk 1]
```diff
@@ -7,6 +7,26 @@
 
 @implementation RNSModalScreen
 
+// When using UIModalPresentationStyleFullScreen the views belonging to the presenting view controller are removed,
+// Thus there is no RCTRootView mounted and no trait observer working.
+// We have to set up an extra observer and send change notifications accoridngly
+// For more information, see https://github.com/software-mansion/react-native-screens/pull/2211.
```
Comment (kkafar) [inline ios/RNSModalScreen.mm:13]:
```suggestion
// When using UIModalPresentationStyleFullScreen the whole view hierarchy mounted under primary `UITransitionView` is removed, including React's root view, which observes for trait collection changes & sends it to `Appearance` module via system notification centre. To workaround this detached-root-view-situation we emit the event to React's `Appearance` ourselves.
// For more information, see https://github.com/software-mansion/react-native-screens/pull/2211.
```

File: ios/RNSModalScreen.mm [hunk 2]
```diff
@@ -7,6 +7,26 @@
...
+- (void)traitCollectionDidChange:(UITraitCollection *)previousTraitCollection
+{
+  [super traitCollectionDidChange:previousTraitCollection];
+  if (RCTSharedApplication().applicationState == UIApplicationStateBackground ||
+      (self.stackPresentation != RNSScreenStackPresentationFullScreenModal)) {
```
Comment (kkafar) [inline ios/RNSModalScreen.mm:18]:
```suggestion
      self.stackPresentation != RNSScreenStackPresentationFullScreenModal) {
```

Idk, why I've put parenthesis there, let's remove them.

File: ios/RNSModalScreen.mm (hunk 1 above)
Comment (kkafar) [reply ios/RNSModalScreen.mm:13]: Let's also add a link to lines from `RCTRootView` file in react native's repository, from where most of this code is taken.

**Review by kkafar — CHANGES_REQUESTED** [review]
very good job!

I've asked for few changes and we're good to go.

File: ios/RNSModalScreen.mm (hunk 1 above)
Comment (alduzy) [reply ios/RNSModalScreen.mm]: [chore: updated comment](https://github.com/software-mansion/react-native-screens/pull/2211/commits/3283fe3aa82504c08449be7abf27df5b0c733512)

**Review by kkafar — APPROVED** [review]
Looks solid now 💪🏻 🎉

**Review by tboba — APPROVED** [review]
LGTM!

**Pull request merged**